### PR TITLE
Validate Sparse Compressed tensor inputs

### PR DESCRIPTION
**Pull request by pearu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #79385

The validation includes regular tensor inputs, batched tensor inputs, as well as hybrid tensor inputs.

### CODE REVIEW

File: aten/src/ATen/SparseCsrTensorUtils.h [hunk 1]
```diff
@@ -181,6 +181,38 @@ inline std::string plainIndicesName(Layout layout) {
...
+      TORCH_CHECK(false, "Not a sparse compressed layout:", layout);
+      return "";
+  }
+}
+
+inline std::string plainDimName(Layout layout) {
+  switch (layout) {
+    case kSparseCsr:
+      return "column";
+    case kSparseCsc:
```
Comment (nikitaved) [inline aten/src/ATen/SparseCsrTensorUtils.h:204]: How is this function different from `compressedDimName`?

File: aten/src/ATen/native/sparse/SparseCsrTensor.cpp [hunk 2]
```diff
@@ -58,199 +59,206 @@ namespace {
...
 } // end anonymous namespace
 
-void _validate_sparse_compressed_tensor_args_worker(const Tensor& compressed_indices, const Tensor& plain_indices, const Tensor& values, const IntArrayRef size, const Layout& layout) {
+/*
+  Validate the arguments to sparse compressed (CSR, CSC, BSR, and BSC)
+  tensor factory functions.
+
+  The CSR and BSR invariants for PyTorch are outlined in
+
+    https://pearu.github.io/csr_tensor_invariants.html
```
Comment (nikitaved) [inline aten/src/ATen/native/sparse/SparseCsrTensor.cpp:68]: It mentions that `int32` is default for CSR, which is not the case for COO. I have also not seen any indexing tensors which are not `int64` in the dense case.

File: aten/src/ATen/native/sparse/SparseCsrTensor.cpp [hunk 3]
```diff
@@ -58,199 +59,206 @@ namespace {
...
-        TORCH_CHECK(
-                    compressed_indices.dim() + 2 <= values.dim(),
-                    "number of dimensions of indices (=", compressed_indices.dim(),
-                    ") plus two must be equal or less than the number of dimensions of values (=", values.dim(), ")");
-        return 2;
-      });
-  int dense_ndim = values.dim() - compressed_indices.dim() - block_ndim;
-  TORCH_CHECK(dense_ndim == 0, "non-zero dense dimensions (=", dense_ndim, ") is not supported for ", layout, " layout");
+              batch_ndim >= 0,
+              compressed_indices_name, " must have dimensionality >= 1 but got ", compressed_indices.dim());
```
Comment (nikitaved) [inline aten/src/ATen/native/sparse/SparseCsrTensor.cpp:113]: Shouldn't the check be `compressed_indices.dim() == batch_ndim + 1`?

File: aten/src/ATen/native/sparse/SparseCsrTensor.cpp [hunk 4]
```diff
@@ -58,199 +59,206 @@ namespace {
...
+    [&] {
+      if (plain_indices.numel() > 0) {
         Tensor compressed_indices_cpu = compressed_indices.to(kCPU);
+        Tensor plain_indices_cpu = plain_indices.to(kCPU);
+        int64_t batch_compressed_stride = compressed_indices_cpu.dim() >= 2 ? compressed_indices_cpu.stride(-2) : 0;
+        int64_t batch_plain_stride = plain_indices_cpu.dim() >= 2 ? plain_indices_cpu.stride(-2) : 0;
         auto compressed_indices_data_ptr = compressed_indices_cpu.data_ptr<index_t>();
-        auto batch_stride = compressed_indices_cpu.dim() >= 2 ? compressed_indices_cpu.stride(-2) : 0;
-        auto compressed_dims = (block_ndim == 0 ? size[compressed_dim] : size[compressed_dim] / blocksize[compressed_dim - batch_ndim]);
+        auto plain_indices_data_ptr = plain_indices_cpu.data_ptr<index_t>();
```
Comment (nikitaved) [inline aten/src/ATen/native/sparse/SparseCsrTensor.cpp:201]: From here and below: it is much more efficient to check the invariance with the TensorIterator, i.e. using `cpu_kernel` and `gpu_kernel`, without memory copies and syncs.

File: aten/src/ATen/native/sparse/SparseCsrTensor.cpp [hunk 5]
```diff
@@ -58,199 +59,206 @@ namespace {
...
+      values.device(),
+      ")");
+  TORCH_CHECK(
+      compressed_indices.get_device() == plain_indices.get_device(),
+      "device of ", compressed_indices_name, " (=",
+      compressed_indices.device(),
+      ") must match device of ", plain_indices_name," (=",
+      plain_indices.device(),
+      ")");
 }
```
Comment (nikitaved) [inline aten/src/ATen/native/sparse/SparseCsrTensor.cpp:262]: A helper function could be used here to eliminate duplicates.

File: aten/src/ATen/native/sparse/SparseCsrTensor.cpp [hunk 6]
```diff
@@ -360,34 +368,54 @@ DimVector _estimate_sparse_compressed_tensor_size(
...
+  const int base_ndim = 2;  // corresponds to compressed and plain indices
+  const int batch_ndim = compressed_indices.dim() - 1;
+  const std::string compressed_indices_name = compressedIndicesName(layout);
+  const std::string plain_indices_name = plainIndicesName(layout);
+  TORCH_CHECK(
+              batch_ndim >= 0,
+              compressed_indices_name, " must have dimensionality >= 1 but got ", compressed_indices.dim());
+  TORCH_CHECK(
+              compressed_indices.dim() == plain_indices.dim(),
+              compressed_indices_name, " and ", plain_indices_name, " dimensionalities must be equal but got ",
```
Comment (nikitaved) [inline aten/src/ATen/native/sparse/SparseCsrTensor.cpp:381]: From here and below: don't we have all these checked in `_validate_sparse_compressed_tensor_args_worker`? Could this method be called before `_validate_sparse_compressed_tensor_args_worker`?

File: test/test_testing.py [hunk 7]
```diff
@@ -1308,7 +1308,7 @@ def test_matching(self):
 
     def test_mismatching_ccol_indices_msg(self):
         actual_ccol_indices = (0, 1, 2)
-        actual_row_indices = (1, 0)
```
Comment (nikitaved) [inline test/test_testing.py:1311]: Otherwise it fails with unsorted indices, right?

File: test/test_sparse_csr.py [hunk 8]
```diff
@@ -171,46 +171,131 @@ def genTensor(self, size, nnz, *, layout, device=None, dtype=torch.float, index_
...
+                if layout is torch.sparse_bsc:
+                    blocksize = tuple(reversed(blocksize))
+
+                if not lst:
+                    return torch.tensor(lst, device=device, dtype=dtype).reshape(0, *blocksize)
+
+                def list_add(lst, value):
+                    if isinstance(lst, list):
+                        return [list_add(item, value) for item in lst]
+                    return lst + value
```
Comment (nikitaved) [inline test/test_sparse_csr.py:223]: Can you have list of lists?

File: test/test_sparse_csr.py [hunk 9]
```diff
@@ -502,6 +574,274 @@ def test_consistency(self, layout, device, dtype, op):
...
+    def test_validate(self, layout, device, dtype):
+        for index_dtype in [torch.int32, torch.int64]:
+            for compressed_indices, plain_indices, values, size in self._generate_small_inputs(
+                    layout, device, dtype, index_dtype, enable_batched=True, enable_hybrid=True):
+                torch._validate_sparse_compressed_tensor_args(compressed_indices, plain_indices, values, size, layout)
+
+    def _generate_invalid_input(self, layout):
+
+        def shape(shape, basedim=0, blocksize=(1, 1)):
+            # Below, we define compressed and plain indices that
```
Comment (nikitaved) [inline test/test_sparse_csr.py:589]: This function is already defined above? Or are they not identical?

**Comment by nikitaved** [comment]
@cpuhrsch , it seems like after this PR is merged, we will be able to construct hybrid compressed tensors. What would be the minimal functionality, when implemented, to be sufficient to say that dense dimensions are supported? `resize` and `to_dense`?

**Comment by cpuhrsch** [comment]
@nikitaved - An extension of `to_sparse_csr` akin to the native function [`to_sparse.sparse_dim`](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/native_functions.yaml#L5879-L5883) so that users can convert from, at least, the strided layout to this layout and a conversion back to strided via `to_dense`. I don't think `resize` is necessary right away, but what makes you suggest it?

**Review by cpuhrsch — COMMENTED** [review]
I think generally this looks good (the large number of changes here are due to the expect files), but @nikitaved I raising good points about various details. Let's wait until Pearu is back from his PTO next week to further process this.

File: aten/src/ATen/native/sparse/SparseCsrTensor.cpp (hunk 2 above)
Comment (pearu) [reply aten/src/ATen/native/sparse/SparseCsrTensor.cpp:68]: https://pearu.github.io/csr_tensor_invariants.html is a little out-of-date: `int64` is the current default for CSR tensors.

The `int32` index support originates from `MKL_INT` being 32-bit integer, see [comment](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/SparseCsrTensorImpl.h#L21-L23).

File: aten/src/ATen/native/sparse/SparseCsrTensor.cpp (hunk 4 above)
Comment (pearu) [reply aten/src/ATen/native/sparse/SparseCsrTensor.cpp:201]: Sure. We should implement this optimization as a follow-up.

In general, validating sparse compressed tensor inputs is (and remains) an expensive operation because one needs to access indices data. Internal tools should use the unsafe constructors for creating such tensors for maximum performance while the correctness of inputs should be checked at the unit tests level.

File: aten/src/ATen/native/sparse/SparseCsrTensor.cpp [hunk 10]
```diff
@@ -58,199 +59,206 @@ namespace {
...
-          for (int i =  1; i <= compressed_dims; i++) {
-            TORCH_CHECK(
-                        compressed_indices_data_ptr[batch_id*batch_stride + i - 1] <= compressed_indices_data_ptr[batch_id*batch_stride + i],
-                        "(Batch element ", batch_id, ") ",
-                        "at position i = ", i, ", the condition ", compressed_indices_name, "[i - 1] <= ", compressed_indices_name, "[i] fails, got ",
-                        compressed_indices_data_ptr[batch_id*batch_stride + i - 1], " <= ", compressed_indices_data_ptr[batch_id*batch_stride + i]);
+          int64_t start = compressed_indices_data_ptr[batch_id*batch_compressed_stride];
+          const std::string at_batch_id = (batch_ndim > 0 ? " at batch id " + std::to_string(batch_id) : "");
+          const std::string batch_indices = (batch_ndim > 0 ? "..., " : "");
+          // 5.1
```
[the remark this answers is not in the document]
Comment (pearu) [reply aten/src/ATen/native/sparse/SparseCsrTensor.cpp:206]: I suggest doing this optimization in a follow-up because this PR targets correctness rather than efficiency, see the general note in https://github.com/pytorch/pytorch/pull/79385#discussion_r907159974

File: test/test_testing.py (hunk 7 above)
Comment (pearu) [reply test/test_testing.py:1311]: Yes, correct. This changeset fixes incorrect test data.

File: test/test_sparse_csr.py (hunk 8 above)
Comment (pearu) [reply test/test_sparse_csr.py:223]: yes, this is exercised in BSR/BSC-related tests.

File: test/test_sparse_csr.py [hunk 11]
```diff
@@ -502,6 +574,274 @@ def test_consistency(self, layout, device, dtype, op):
...
+                if layout is torch.sparse_bsc:
+                    blocksize = tuple(reversed(blocksize))
+
+                if not lst:
+                    return torch.tensor(lst, device=device, dtype=dtype).reshape(0, *blocksize)
+
+                def list_add(lst, value):
+                    if isinstance(lst, list):
+                        return [list_add(item, value) for item in lst]
+                    return lst + value
```
[the remark this answers is not in the document]
Comment (pearu) [reply test/test_sparse_csr.py:625]: Fixed: introduced `_generate_small_inputs_utils` method that returns `shape` and `values` utility functions.

File: test/test_sparse_csr.py (hunk 9 above)
Comment (pearu) [reply test/test_sparse_csr.py:589]: Fixed via https://github.com/pytorch/pytorch/pull/79385#discussion_r907200124

File: aten/src/ATen/native/sparse/SparseCsrTensor.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (nikitaved) [reply aten/src/ATen/native/sparse/SparseCsrTensor.cpp:206]: I agree, and I am currently working on it.

File: aten/src/ATen/native/sparse/SparseCsrTensor.cpp (hunk 6 above)
Comment (pearu) [reply aten/src/ATen/native/sparse/SparseCsrTensor.cpp:381]: Yes, these checks are run twice when a user does not specify the tensor size: in `_estimate_sparse_compressed_tensor_size` followed by `_validate_sparse_compressed_tensor_args_worker` call. However, we'll need these checks here to make sure that the inputs to `_estimate_sparse_compressed_tensor_size` are valid similar to `_validate_sparse_compressed_tensor_args_worker` that assumes a specified size. Notice that the overhead of these checks is minimal (it is about comparing two integers), so, performance-wise, it should be ok.

No, `_estimate_sparse_compressed_tensor_size` cannot be called before `_validate_sparse_compressed_tensor_args_worker` in general because (i) when an user specifies the `size`, there is no need to estimate the size from inputs, and (ii) the sizes returned by `_estimate_sparse_compressed_tensor_size` and specified by the user may be different while both being valid sizes (`_estimate_sparse_compressed_tensor_size` returns a *minimal* valid size for the given input).

File: aten/src/ATen/native/sparse/SparseCsrTensor.cpp (hunk 5 above)
Comment (pearu) [reply aten/src/ATen/native/sparse/SparseCsrTensor.cpp:262]: I am not sure if a helper function is worth implementing here considering that it would be the only use case of such a function and the outputs of these checks are fully tested in
https://github.com/pytorch/pytorch/blob/384ae783c478040bd9ec5142e648786b10598a52/test/test_sparse_csr.py#L741-L759
What do you think, @nikitaved?

Comment (nikitaved) [reply aten/src/ATen/native/sparse/SparseCsrTensor.cpp:262]: Makes sense. Fine by me :)

**Review by nikitaved — APPROVED** [review]
Thank you, @pearu ! LGTM.

**Review by cpuhrsch — APPROVED** [review]
